okhttp/src/main/java/io/grpc/okhttp/OkHttpProtocolNegotiator.java

```diff
@@ -152,21 +212,68 @@ public String negotiate(SSLSocket sslSocket, String hostname, List<Protocol> pro
     /**
      * Override {@link Platform}'s configureTlsExtensions for Android older than 5.0, since OkHttp
      * (2.3+) only support such function for Android 5.0+.
+     *
+     * <p>Note: Prior to Android Q, the standard way of accessing some Conscrypt features was to
+     * use reflection to call hidden APIs. Beginning in Q, there is public API for all of these
+     * features. We attempt to use the public API where possible. Otherwise, fall back to use the
+     * old reflective API.
      */
     @Override
     protected void configureTlsExtensions(
         SSLSocket sslSocket, String hostname, List<Protocol> protocols) {
+      SSLParameters sslParams = sslSocket.getSSLParameters();
       // Enable SNI and session tickets.
       if (hostname != null) {
-        SET_USE_SESSION_TICKETS.invokeOptionalWithoutCheckedException(sslSocket, true);
-        SET_HOSTNAME.invokeOptionalWithoutCheckedException(sslSocket, hostname);
+        try {
+          if (SSL_SOCKETS_IS_SUPPORTED_SOCKET != null
+              && (boolean) SSL_SOCKETS_IS_SUPPORTED_SOCKET.invoke(null, sslSocket)) {
+            SSL_SOCKETS_SET_USE_SESSION_TICKET.invoke(null, sslSocket, true);
+          } else {
+            SET_USE_SESSION_TICKETS.invokeOptionalWithoutCheckedException(sslSocket, true);
+          }
+        } catch (IllegalAccessException e) {
+          throw new RuntimeException(e);
+        } catch (InvocationTargetException e) {
+          throw new RuntimeException(e);
+        }
+
+        try {
+          if (SNI_HOST_NAME != null) {
+            SET_SERVER_NAMES
+                .invokeOptionalWithoutCheckedException(
+                    sslParams,
+                    Collections.singletonList(SNI_HOST_NAME.newInstance(hostname)));
+          } else {
+            SET_HOSTNAME.invokeOptionalWithoutCheckedException(sslSocket, hostname);
+          }
+        } catch (IllegalAccessException e) {
+          throw new RuntimeException(e);
+        } catch (InstantiationException e) {
+          throw new RuntimeException(e);
+        } catch (InvocationTargetException e) {
+          throw new RuntimeException(e);
+        }
+      }
+
+      // Enable ALPN and NPN if necessary.
+      String[] protocolNames = protocolIds(protocols);
+      SET_APPLICATION_PROTOCOLS
+          .invokeOptionalWithoutCheckedException(sslParams, (Object) protocolIds(protocols));
+      sslSocket.setSSLParameters(sslParams);
+      // Check protocol configuration through SSLParameters succeeds. If not, fall back to
+      // configure with the old reflective method.
+      // Workaround for Conscrypt issue: https://github.com/google/conscrypt/issues/832
```

Comment:
As written, this is not working around Conscrypt. This is checking whether the API is supported in a Conscrypt-compatible way. For this to be a workaround we would need to do the does-GET_APPLICATION_PROTOCOL-throw check before this, or we would need to say "we aren't checking getApplicationProtocol as a Conscrypt workaround".

Reply:
I guess you are really referring to `SSLParameters#getApplicationProtocols()` instead of `SSLSocket#getApplicationProtocol()` here. The former never throw (we aren't doing the latter during configuration). The `getApplicationProtocols()` check really is to see if protocols configured on `SSLParameters` are propagated to `SSLSocket`, which is only because Conscrypt on Android does this incorrectly. Normally we wouldn't need to do this. Isn't this a workaround?

Reply:
I was referring to SSLSocket.getApplicationProtocol().

An API was added in Java 9 and Android API 27 to configure ALPN. But that API is implemented by multiple implementations. So we can't assume that if the API is there that it works. I was suggesting to use `SSLSocket.getApplicationProtocol()` before any negotiation has occurred to see if ALPN is supported in this fashion. If the method throws `UnsupportedOperationException` then the implementation does not support that API and we should try other APIs. That is the way we [decided to handle it in Netty](https://github.com/netty/netty/blob/6c05d16967df5cadcebcc498223eec3f25bacf89/handler/src/main/java/io/netty/handler/ssl/Java9SslUtils.java#L67).

Reply:
I see. That was what we used to find the Conscrypt bug. Due to that bug, even the invocation of `SSLSocket.getApplicationProtocol()` before any negotiation has occurred does not throw `UnsupportedOperationException`, ALPN is still not configured correctly. Here, we are working around that bug (only for versions earlier than R, aka, FileDescriptorSocket). Consequently, this check also helps us verifying the SSLSocket is configured with ALPN, isn't it? So do we still need the does-GET_APPLICATION_PROTOCOL-throw check?